smolyay/basis.py

```diff
@@ -403,29 +478,312 @@ def _derivative(self, x):
+        elif basis_functions is None:
+            basis_functions = []
+
+        if len(basis_functions) > 1:
+            ref_domain = basis_functions[0].domain
+            ref_type = type(basis_functions[0])
+            if any(
+                not numpy.array_equal(b.domain, ref_domain)
+                or type(b) is not ref_type
+                for b in basis_functions[1:]
+            ):
+                raise TypeError(
+                    "Basis functions must be be BasisFunction objects with the same domain and typing."
+                )
+        self._basis_functions = list(basis_functions)
 
     @property
     def basis_functions(self):
         """list: Basis functions."""
         return self._basis_functions
+
+    @property
+    def domain(self):
+        """numpy.ndarray: Domain of the `basis_functions`"""
+        if len(self._basis_functions) == 0:
+            raise AttributeError("No basis functions to derive a domain.")
+        return self._basis_functions[0].domain
+
+    def __len__(self):
+        return len(self._basis_functions)
+
+    def __getitem__(self, key):
+        return self.basis_functions[key]
+
+    def scale_to_domain(self, points, old_domain):
+        if len(self._basis_functions) == 0:
+            raise AttributeError("No basis functions to derive a domain.")
+        return self.basis_functions[0].scale_to_domain(points, old_domain)
+
+    def __call__(self, X, domain=None):
+        """Evaluate all the basis functions in the set
+
+        Calls all the basis function(s) at index and evaluates at X.
+
+        Parameters
+        ----------
+        X : array-like
+            the points to evaluate.
+
+        domain : numpy array of shape (2,)
+            the lower and upper bounds of X.
+
+        Returns
+        -------
+        scalar or ndarray
+            the values of the basis functions.
+        """
+        if not domain is None:
+            new_X = self.scale_to_domain(X, domain)
+        else:
+            new_X = numpy.asarray(X)
+        y = numpy.zeros(
+            [len(self)] + list(new_X.shape),
+            dtype=complex if any(bf._is_complex for bf in self) else float,
+        )
+        for i, bf in enumerate(self._basis_functions):
+            y[i] = bf(new_X)
+        return y
+
+    def derivative(self, X, domain, n=1):
+        """Evaluate all the derivative of basis functions in the set
+
+        Calls the derivative for all the basis function(s) and
+        evaluates at X.
+
+        Parameters
+        ----------
+        X : array-like
+            the points to evaluate.
+
+        domain : numpy array of shape (2,)
+            the lower and upper bounds of X.
+
+        n : int, optional
+            order of derivative. Default is 1.
+
+        Returns
+        -------
+        scalar or ndarray
+            the values of the basis functions.
+        """
+        new_X = self.scale_to_domain(numpy.asarray(X), domain)
```

Comment:
This method should take care of converting if needed.
```suggestion
        new_X = self.scale_to_domain(X, domain)
```
You'll need the `if-else` to allow `domain=None` as the default.